packages/app-metadata-controller/src/AppMetadataController.ts

```diff
@@ -0,0 +1,195 @@
+  (): AppMetadataControllerState => ({
+    currentAppVersion: '',
+    previousAppVersion: '',
+    previousMigrationVersion: 0,
+    currentMigrationVersion: 0,
+  });
+
+/**
+ * Returns the state of the {@link AppMetadataController}.
+ */
+export type AppMetadataControllerGetStateAction = ControllerGetStateAction<
+  typeof controllerName,
+  AppMetadataControllerState
+>;
+
+/**
+ * Actions exposed by the {@link AppMetadataController}.
+ */
+export type AppMetadataControllerActions = AppMetadataControllerGetStateAction;
+
+/**
+ * Event emitted when the state of the {@link AppMetadataController} changes.
+ */
+export type AppMetadataControllerStateChangeEvent = ControllerStateChangeEvent<
+  typeof controllerName,
+  AppMetadataControllerState
+>;
+
+/**
+ * Events that can be emitted by the {@link AppMetadataController}
+ */
+export type AppMetadataControllerEvents = AppMetadataControllerStateChangeEvent;
+
+/**
+ * Actions that this controller is allowed to call.
+ * Currently set to never as this controller doesn't call any other controllers.
+ */
+type AllowedActions = never;
+
+/**
+ * Events that this controller is allowed to subscribe.
+ */
+type AllowedEvents = never;
+
+/**
+ * Messenger type for the {@link AppMetadataController}.
+ *
+ * @returns A restricted messenger type that defines the allowed actions and events
+ * for the AppMetadataController
+ */
+export type AppMetadataControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  AppMetadataControllerActions | AllowedActions,
+  AppMetadataControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * Metadata configuration for the {@link AppMetadataController}.
+ *
+ * Defines persistence and anonymity settings for each state property.
+ */
+const controllerMetadata = {
+  currentAppVersion: {
+    persist: true,
+    anonymous: true,
+  },
+  previousAppVersion: {
+    persist: true,
+    anonymous: true,
+  },
+  previousMigrationVersion: {
+    persist: true,
+    anonymous: true,
+  },
+  currentMigrationVersion: {
+    persist: true,
+    anonymous: true,
+  },
+} satisfies StateMetadata<AppMetadataControllerState>;
+
+/**
+ * The AppMetadata controller stores metadata about the current extension instance,
+ * including the currently and previously installed versions, and the most recently
+ * run migration.
+ *
+ */
+export class AppMetadataController extends BaseController<
+  typeof controllerName,
+  AppMetadataControllerState,
+  AppMetadataControllerMessenger
+> {
+  /**
+   * Constructs a AppMetadata controller.
+   *
+   * @param options - the controller options
+   * @param options.state - Initial controller state.
+   * @param options.messenger - Messenger used to communicate with BaseV2 controller.
+   * @param options.currentMigrationVersion - The migration version to store in state.
+   * @param options.currentAppVersion - The app version to store in state.
+   */
+  constructor({
+    state = {},
```

Comment:
It looks like the scenario where the constructor is initialized with no explicit state is not tested, which is why CI is failing. I've suggested an update you can add in `AppMetadataController.test.ts` to resolve this.